File: problems/grass/grass.py
from math import sqrt

while True:
    try:
        n, l, w = map(int, input().split())
        intervals = []
        for i in range(n):
            x, r = map(int, input().split())
            if r > w / 2:
                interval_width = sqrt(r ** 2 - (w ** 2 / 4))
                intervals.append((x - interval_width, x + interval_width))
        intervals.sort()

        end = 0
        i = 0
        res = 0
        while True:
            new_end = end
            while i < len(intervals) and intervals[i][0] <= end:
                new_end = max(new_end, intervals[i][1])
                i += 1
            res += 1
            if new_end >= l:
                print(res)
                break
            if new_end == end:
                print(-1)
                break
            end = new_end
    except EOFError:
        exit()

File: problems/grass/grass.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
typedef pair<double, double> pdd;

int main() {
    ll n, l, w;
    while (cin >> n >> l >> w) {
        vector<pdd> intervals;
        intervals.reserve(n);
        for (int i = 0; i < n; ++i) {
            ll x, r;
            cin >> x >> r;
            if (2 * r > w) {
                double iw = sqrt(r * r - double(w) * w / 4.);
                intervals.emplace_back(x - iw, x + iw);
            }
        }
        sort(intervals.begin(), intervals.end());

        ll res = 0, i = 0;
        double end = 0;
        while (true) {
            double new_end = end;
            while (i < intervals.size() && intervals[i].first <= end) {
                new_end = max(new_end, intervals[i].second);
                i += 1;
            }
            res += 1;
            if (new_end == end) {
                cout << -1 << "\n";
                break;
            } else if (new_end >= l) {
                cout << res << "\n";
                break;
            }
            end = new_end;
        }
    }
}